### ISSUE TITLE
[Question] Eagle3 training seems to freeze target embeddings but train a new draft lm_head

### ISSUE THREAD

**Issue by Hannibal046** [sgl-project/SpecForge#478]
Hi, thanks for open-sourcing this project.

While reading the Eagle3 training code, I noticed a behavior that seems different from my understanding of the Eagle3 paper / method, and I wanted to confirm whether this is intentional.

## What I observed

In the current Eagle3 training path:

1. The draft model loads the **target model embeddings**:
   - `scripts/train_eagle3.py`
     ```python
     draft_model.load_embedding(args.target_model_path, embedding_key=args.embedding_key)
     draft_model.freeze_embedding()
     ```
   - so the draft embedding is copied from the target model and then frozen.

2. The draft model defines its **own lm_head**:
   - `specforge/modeling/draft/llama3_eagle.py`
     ```python
     self.lm_head = nn.Linear(config.hidden_size, config.draft_vocab_size, bias=False)
     ```
   - and logits are produced from this draft `lm_head`:
     ```python
     return self.lm_head(norm_hidden_states)
     ```

This seems a bit surprising to me, especially since the embedding is reused from the target model, but the output head is not. Is this behavior intentional?

Thanks.

**Comment by 0zzx0** [comment]
Same issue here. This seems different from the original paper.

**Comment by jiapingW** [comment]
Because the draft model uses a separate vocabulary, which is a subset of the target model's vocabulary, and since all tokens are in the target model's vocabulary, the target model's embedding needs to be used. However, when generating tokens, the draft model's own vocabulary needs to be used to calculate the loss, so lm_head needs to be retrained.

**Comment by Hannibal046** [comment]
@jiapingW Hi, thanks for the answer! Since the `lm_head` is essentially just |V| vectors of dimension D used to calculate the inner product with hidden states to get logits, wouldn't it also be viable to just select the corresponding vectors from the target model's `lm_head` (whose IDs are in the new vocabulary)? This approach seems like it would work without requiring any retraining.

**Comment by jiapingW** [comment]
We follow the impl of [eagle3](https://github.com/SafeAILab/EAGLE/blob/main/eagle/traineagle3/cnets.py#L531). You opinion is reasonable, we try it on dflash but no performance difference was observed.

**Comment by Hannibal046** [comment]
Get! Thanks for the reply!
